internal/server/server.go

```diff
@@ -0,0 +1,135 @@
+// SPDX-FileCopyrightText: 2025 The Kepler Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/exporter-toolkit/web"
+	"github.com/sustainable-computing-io/kepler/internal/service"
+)
+
+// APIService defines the interface for the HTTP server providing API endpoints
+type APIService interface {
+	service.Service
+	Register(endpoint, summary, description string, handler http.Handler) error
+}
+
+// APIServer implements APIServer
+type APIServer struct {
+	// input
+	logger      *slog.Logger
+	listenAddrs []string
+
+	// http
+	server              *http.Server
+	mux                 *http.ServeMux
+	endpointDescription string
+}
+
+var _ APIService = (*APIServer)(nil)
+
+type Opts struct {
+	logger      *slog.Logger
+	listenAddrs []string
+}
+
+// OptionFn is a function sets one more more options in Opts struct
+type OptionFn func(*Opts)
+
+// WithLogger sets the logger for the APIServer
+func WithLogger(logger *slog.Logger) OptionFn {
+	return func(o *Opts) {
+		o.logger = logger
+	}
+}
+
+// WithListenAddress sets the listening addresses for the APIServer
+func WithListenAddress(addr []string) OptionFn {
+	return func(o *Opts) {
+		o.listenAddrs = addr
+	}
+}
+
+// DefaultOpts returns the default options
+func DefaultOpts() Opts {
+	return Opts{
+		logger:      slog.Default(),
+		listenAddrs: []string{":28282"}, // Default HTTP Port
+	}
+}
+
+// NewAPIServer creates a new HTTPAPIServer instance
+func NewAPIServer(applyOpts ...OptionFn) *APIServer {
+	opts := DefaultOpts()
+	for _, apply := range applyOpts {
+		apply(&opts)
+	}
+
+	mux := http.NewServeMux()
+	server := &http.Server{
+		Handler: mux,
+	}
+	apiServer := &APIServer{
+		logger:      opts.logger.With("service", "api-server"),
+		listenAddrs: opts.listenAddrs,
+		mux:         mux,
+		server:      server,
+	}
+
+	return apiServer
+}
+
+func (s *APIServer) Name() string {
+	return "api-server"
+}
+
+func (s *APIServer) Start(ctx context.Context) error {
+	s.logger.Info("Starting HTTP server", "listening-on", s.listenAddrs)
+	if len(s.listenAddrs) == 0 {
+		return fmt.Errorf("no listening address provided")
+	}
+
+	errCh := make(chan error)
+	go func() {
+		webCfg := &web.FlagConfig{
+			WebListenAddresses: &s.listenAddrs,
+			WebConfigFile:      pointer(""), // TODO: support passing config file
+		}
+		errCh <- web.ListenAndServe(s.server, webCfg, s.logger)
+	}()
```

Comment:
suggestion: service `Start()` already runs in a separate go-routine, here we are creating another. if possible please simplify.

Reply:
there are 2 things
1. watch for context (supplied from main) to be cancelled and quit
2. or end when ListenAndServe() exits

both are blocking and thus we need a go routine for the server